plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Stablecoin.hs

```diff
@@ -0,0 +1,385 @@
+initialState StateMachineClient{scInstance=StateMachineInstance{validatorInstance}} =
+    BankState
+        { bsReserves = 0
+        , bsStablecoins = 0
+        , bsReservecoins = 0
+        , bsForgingPolicyScript = monetaryPolicyHash $ forwardingMPS $ scriptHash validatorInstance
+        }
+
+{-# INLINEABLE convert #-}
+-- | Convert peg currency units to base currency units using the
+--   observed conversion rate
+convert :: ConversionRate -> PC (Ratio Integer) -> BC (Ratio Integer)
+convert rate (PC pc) =
+    BC $ rate * pc
+
+{-# INLINEABLE liabilities #-}
+-- | The bank's liabilities (total value of stablecoins in base currency)
+liabilities ::
+    BankState
+    -> ConversionRate
+    -> BC (Ratio Integer)
+liabilities BankState{bsReserves=BC reserves,bsStablecoins=SC stablecoins} cr =
+    let BC stableCoinLiabilities = convert cr (PC $ fromInteger stablecoins)
+    in BC (min (fromInteger reserves) stableCoinLiabilities)
+
+{-# INLINEABLE equity #-}
+-- | The bank's equity (what's left of the reserves after subtracting
+--   liabilities).
+equity ::
+    BankState
+    -> ConversionRate
+    -> BC (Ratio Integer)
+equity r@BankState{bsReserves=BC reserves} cr =
+    let BC l = liabilities r cr
+    in BC (fromInteger reserves - l)
+
+-- | Stablecoin parameters.
+data Stablecoin =
+    Stablecoin
+        { scOracle                  :: PubKey -- ^ Public key of the oracle that provides exchange rates
+        , scFee                     :: Ratio Integer -- ^ Fee charged by bank for transactions. Calculated as a fraction of the total transaction volume in base currency.
+        , scMinReserveRatio         :: Ratio Integer -- ^ The minimum ratio of reserves to liabilities
+        , scReservecoinDefaultPrice :: BC Integer -- ^ The price of a single reservecoin if no reservecoins have been issued
+        , scBaseCurrency            :: (CurrencySymbol, TokenName) -- ^ The base currency. Value of this currency will be locked by the stablecoin state machine instance
+        , scStablecoinTokenName     :: TokenName -- ^ 'TokenName' of the stablecoin
+        , scReservecoinTokenName    :: TokenName -- ^ 'TokenName' of the reservecoin
+        }
+    deriving stock (Generic, Haskell.Eq, Haskell.Show)
+    deriving anyclass (ToJSON, FromJSON)
+
+{-# INLINEABLE minReserve #-}
+-- | Minimum number of base currency coins held by the bank
+minReserve :: Stablecoin -> ConversionRate -> BankState -> BC (Ratio Integer)
+minReserve Stablecoin{scMinReserveRatio} cr BankState{bsStablecoins=SC sc} =
+    let BC r = convert cr (PC $ fromInteger sc)
+    in BC $ scMinReserveRatio * r
+
+{-# INLINEABLE reservecoinNominalPrice #-}
+-- | Price of a single reservecoin in base currency
+reservecoinNominalPrice :: Stablecoin -> BankState -> ConversionRate -> BC (Ratio Integer)
+reservecoinNominalPrice Stablecoin{scReservecoinDefaultPrice} bankState@BankState{bsReservecoins=RC rc} cr =
+    if rc /= 0
```

Comment:
Hm, maybe we need a `Natural` type...

Reply:
I'd probably put this in a guard. Literal patterns don't work but guards do, e.g. `foo x | x == 0 = ...`